Fix flaky test_udf_remote_message_delay_timeout_to_self

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41217 Fix flaky test_udf_remote_message_delay_timeout_to_self**

Closes https://github.com/pytorch/pytorch/issues/41019. Due to the possibility of callback
finishCreatingOwnerRRef running after request_callback has processed and
created the owner RRef, we could actually end up with 0 owners on the node,
since the callback removes from the owners_ map. In this case, shutdown is fine
since there are no owners. On the other hand, if the callback runs first, there
will be 1 owner which we will delete in shutdown when we detect it has no
forks. So either way, shutdown works fine and we don't need to enforce there to
be 1 owner.

Differential Revision: [D22469806](https://our.internmc.facebook.com/intern/diff/D22469806/)

rohan-varma committed:
Fix flaky test_udf_remote_message_delay_timeout_to_self

Fixes this flaky test. Due to the possibility of callback
finishCreatingOwnerRRef running after request_callback has processed and
created the owner RRef, we could actually end up with 0 owners on the node,
since the callback removes from the owners_ map. In this case, shutdown is fine
since there are no owners. On the other hand, if the callback runs first, there
will be 1 owner which we will delete in shutdown when we detect it has no
forks. So either way, shutdown works fine and we don't need to enforce there to
be 1 owner.

Differential Revision: [D22469806](https://our.internmc.facebook.com/intern/diff/D22469806/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3551,16 +3551,14 @@ def _test_remote_message_delay_timeout(self, func, args, dst=None):
             wait_until_owners_and_forks_on_rank(2, 2, rank=dst_rank)
-        else:
-            # 1 owner, fork deleted
-            wait_until_owners_and_forks_on_rank(1, 0, rank=dst_rank)
```
Question, by "the `finishCreatingOwnerRRef` callback would run before the `request_callback_impl` finishes processing the request", the request here is not the original `rpc.remote()` request, instead it is the request sent by `wait_until_owners_and_forks_on_rank`? Is this correct?

rohan-varma replied:
It's actually the `rpc.remote()` request. What happens in this test is that rank0 sends `rpc.remote()` to rank0, but we inject failure so the future times out in 1ms. Since the future is marked with an error, callbacks can be run at any time by the thread that set the errors. 

In another thread, the rpc.remote() is still being processed (we marked it with an error but that does not stop the processing on the receiver). There are 2 cases:

1. Callbacks are run before the request is processed. In this case, we call `finishCreatingOwnerRRef` and remove the owner. Then, rpc.remote() sees that the owner does not exist and creates a new `ownerRRef`. Then there will be 1 owner on the node (this is what the test previously assumed would happen)

2. Callbacks are run after the request is processed. In this case, rpc.remote() used the existing ownerRRef, and in the callback we will remove it, so there will be 0 owners (this fix should make this case work).

mrshenli replied:
I see. I wonder if we should handle this error more explicitly. For the rpc-to-self case, if the receiving end expects the `OwnerRRef` to be there before processing the request, can we change the following two lines and guard them by checking `rrefId != forkId`? If it is rpc-to-self, and the OwnerRRef does not exist, shall we just let it crash?

https://github.com/pytorch/pytorch/blob/53af9df557aff745edf24193ece784fd008c6f19/torch/csrc/distributed/rpc/request_callback_impl.cpp#L318

https://github.com/pytorch/pytorch/blob/53af9df557aff745edf24193ece784fd008c6f19/torch/csrc/distributed/rpc/request_callback_impl.cpp#L429

rohan-varma replied:
Added this in the latest version.

mrshenli reviewed (approved):
Thanks for fixing, the change LGTM (if I comment inline is correct).

rohan-varma committed:
Update on "Fix flaky test_udf_remote_message_delay_timeout_to_self"


Closes https://github.com/pytorch/pytorch/issues/41019. Due to the possibility of callback
finishCreatingOwnerRRef running after request_callback has processed and
created the owner RRef, we could actually end up with 0 owners on the node,
since the callback removes from the owners_ map. In this case, shutdown is fine
since there are no owners. On the other hand, if the callback runs first, there
will be 1 owner which we will delete in shutdown when we detect it has no
forks. So either way, shutdown works fine and we don't need to enforce there to
be 1 owner.

Differential Revision: [D22469806](https://our.internmc.facebook.com/intern/diff/D22469806/)

[ghstack-poisoned]

rohan-varma commented:
@mrshenli, Updated based on our discussion to enforce that the ownerRRef should always exist in the map if it's an RPC to self, and otherwise, we just let it crash. This results in a crash in `request_callback`, In our test case, since the future is already marked with the timeout error, the test still passes as expected.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -315,7 +315,14 @@ void RequestCallbackImpl::processRpc(
+      if (rrefId == forkId) {
+        // Creating an owner RRef on self, should already exist in owners map
+        ownerRRef =
+            ctx.getOwnerRRef(rrefId, /* forceCreated */ true)->constValue();
```
Just confirming, will the exception thrown by `TORCH_INTERNAL_ASSERT` in `getOwnerRRef` be captured by the try-catch block in `RequestCallbackImpl::processMessage`, so that we don't lose a thread?

rohan-varma replied:
Yep, this error is indeed caught and processed by `RequestCallback::handleError` which logs the error and sets it on the future, and the thread doesn't crash. We do see the error and stacktrace logged such as:

```
> E0715 13:38:22.403939 1125458 request_callback_impl.cpp:168] Received error while processing request type 5: false INTERNAL ASSERT FAILED at "ca
ffe2/torch/csrc/distributed/rpc/rref_context.cpp":379, please report a bug to PyTorch. Expected OwnerRRef with id GloballyUniqueId(created_on=0, l
ocal_id=0) to be created.
```

mrshenli reviewed (approved):
LGTM! (Assuming we don't lose a thread on crash)

rohan-varma committed:
Update on "Fix flaky test_udf_remote_message_delay_timeout_to_self"


Closes https://github.com/pytorch/pytorch/issues/41019. Due to the possibility of callback
finishCreatingOwnerRRef running after request_callback has processed and
created the owner RRef, we could actually end up with 0 owners on the node,
since the callback removes from the owners_ map. In this case, shutdown is fine
since there are no owners. On the other hand, if the callback runs first, there
will be 1 owner which we will delete in shutdown when we detect it has no
forks. So either way, shutdown works fine and we don't need to enforce there to
be 1 owner.

Differential Revision: [D22469806](https://our.internmc.facebook.com/intern/diff/D22469806/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix flaky test_udf_remote_message_delay_timeout_to_self"


Closes https://github.com/pytorch/pytorch/issues/41019. Due to the possibility of callback
finishCreatingOwnerRRef running after request_callback has processed and
created the owner RRef, we could actually end up with 0 owners on the node,
since the callback removes from the owners_ map. In this case, shutdown is fine
since there are no owners. On the other hand, if the callback runs first, there
will be 1 owner which we will delete in shutdown when we detect it has no
forks. So either way, shutdown works fine and we don't need to enforce there to
be 1 owner.

Differential Revision: [D22469806](https://our.internmc.facebook.com/intern/diff/D22469806/)

[ghstack-poisoned]
